examples/checkpointing/checkpointing_example.py

```diff
@@ -1,141 +1,56 @@
-    u.data[:] = 0
-    m0 = smooth10(model.m.data, model.shape_domain)
-    dm = np.float32(model.m.data - m0)
-    return fw, gradop, u, rec_s, m0, src, rec_g, v, grad, rec_t, dm, nt, dt
-
-
-def gradient(fw, gradop, u, maxmem, rec_s, m0, src, rec_g, v, grad, rec_t, nt, dt):
-    cp = DevitoCheckpoint([u])
-    n_checkpoints = None
-    if maxmem is not None:
-        n_checkpoints = int(floor(maxmem*10**6/(cp.size*u.data.itemsize)))
-
-    wrap_fw = CheckpointOperator(fw, {'u': u, 'rec': rec_s, 'm': m0, 'src': src,
-                                      'dt': dt})
-    wrap_rev = CheckpointOperator(gradop, {'u': u, 'v': v, 'm': m0, 'rec': rec_g,
-                                           'grad': grad, 'dt': dt})
-    wrp = Revolver(cp, wrap_fw, wrap_rev, n_checkpoints, nt-2)
-
-    wrp.apply_forward()
-
-    rec_g.data[:] = rec_s.data[:] - rec_t.data[:]
-
-    wrp.apply_reverse()
-
-    # The result is in grad
-    return grad.data
-
+    @cached_property
+    def forward_operator(self):
+        # Verify uses a forward operator with save=False
+        return self.verify_operator
 
-def verify(gradient, dm, m0, u, rec_s, fw, src, rec_t, dt):
-    # Objective function value
-    F0 = .5*linalg.norm(rec_s.data - rec_t.data)**2
-    # <J^T \delta d, dm>
-    G = np.dot(gradient.reshape(-1), dm.reshape(-1))
-    # FWI Gradient test
-    H = [0.5, 0.25, .125, 0.0625, 0.0312, 0.015625, 0.0078125]
-    error1 = np.zeros(7)
-    error2 = np.zeros(7)
+    @cached_property
+    def gradient_operator(self):
+        return GradientOperator(self.model, self.src, self.rec_g, time_order=self.time_order,
+                              spc_order=self.space_order, save=False)
 
-    for i in range(0, 7):
-        # Add the perturbation to the model
-        mloc = m0 + H[i] * dm
-        # Set field to zero (we're re-using it)
-        u.data.fill(0)
-        rec_s.data.fill(0)
-        # Receiver data for the new model
-        # Results will be in rec_s
-        fw.apply(u=u, rec=rec_s, m=mloc, src=src, dt=dt)
-        d = rec_s.data
-        # First order error Phi(m0+dm) - Phi(m0)
-        error1[i] = np.absolute(.5*linalg.norm(d - rec_t.data)**2 - F0)
-        # Second order term r Phi(m0+dm) - Phi(m0) - <J(m0)^T \delta d, dm>
-        error2[i] = np.absolute(.5*linalg.norm(d - rec_t.data)**2 - F0 - H[i] * G)
+    def gradient(self, m0, maxmem=None):
+        cp = DevitoCheckpoint([self.forward_field])
+        n_checkpoints = None
+        if maxmem is not None:
+            n_checkpoints = int(floor(maxmem*10**6/(cp.size*self.forward_field.data.itemsize)))
 
-    # Test slope of the  tests
-    p1 = np.polyfit(np.log10(H), np.log10(error1), 1)
-    p2 = np.polyfit(np.log10(H), np.log10(error2), 1)
-    assert np.isclose(p1[0], 1.0, rtol=0.1)
-    assert np.isclose(p2[0], 2.0, rtol=0.1)
+        wrap_fw = CheckpointOperator(self.forward_operator, {'u': self.forward_field, 'rec': self.rec, 'm': m0, 'src': self.src, 'dt': self.dt})
```

Comment:
standard kwargs format ``a=b`` rather than ``{'a': b}`` would be better

Reply:
Fixed. Thanks for this one.